File: catfish-ui/src/components/shared/form-models/formData.ts

/**
 * The Form Data model which carries the submission data of a form.
 * */

import { Guid } from "guid-typescript"
import { eState } from "../constants";

import { FieldData } from './fieldData'

export interface FormData {
    /**
     * Unique form-data object ID.
     * */
    id: Guid;

    /**
     * The ID of the form
     * */
    formId: Guid

    /**
     * Created timestamp
     * */
    created: Date;

    /**
     * Last updated timestamp
     * */
    updated: Date;

    /**
     * List of field-data objects belongs to this form-data object. Each field-data
     * instance holds data submitted for a field in the parent form.
     * */
    fieldData: FieldData[];

    state: eState


}

File: Catfish/Catfish.API.Repository/Models/Forms/FormData.cs
namespace Catfish.API.Repository.Models.Forms
{
    public class FormData
    {
        /// <summary>
        /// Unique FormData object ID
        /// </summary>
        public Guid Id { get; set; }

        /// <summary>
        /// Form definition ID
        /// </summary>
        public Guid FormId { get; set; }

        /// <summary>
        /// Created timestamp.
        /// </summary>
        public DateTime Created { get; set; } = DateTime.Now;

        /// <summary>
        /// Last updated timestamp.
        /// </summary>
        public DateTime Updated { get; set; }

        /// <summary>
        /// List of fields associated with the form in JSON serialized form-data object.
        /// </summary>
        [JsonIgnore]
        public string? SerializedFieldData { get; set; }

        /// <summary>
        /// List of field data objects associated with the form-data object.
        /// </summary>
        [NotMapped]
        public IList<object>? FieldData
        {
            get => SerializedFieldData == null ? null : JsonConvert.DeserializeObject<List<object>>(SerializedFieldData);
            set => SerializedFieldData = value == null ? null : JsonConvert.SerializeObject(value);
        }

        public eState State { get; set; }
    }
}
